system/datatypes/mailinglist.php

```diff
@@ -1018,18 +1073,31 @@ protected function mailingSendAction(HyphaRequest $request) {
 		// iterate over the confirmed addresses and send the mailing
 		$xpath = './/' . self::FIELD_NAME_ADDRESS . '[@' . self::FIELD_NAME_STATUS . '="' . self::ADDRESS_STATUS_CONFIRMED . '"]';
 		$receivers = $this->findAddresses($xpath);
+		$receivers = [1,2,3,4,5,6,7,8,9,10]; //TODO for testing purposes
+		if (count($receivers) === 0) {
+			notify('error', __('ml-mail-has-invalid-status')); //TODO other error text?
+			return null;
+		}
+
 		$linkToMailing = $this->path(self::PATH_MAILS_VIEW_ID, ['id' => $mailingId]);
 		/** @var HyphaDomElement $receiver */
 		foreach ($receivers as $receiver) {
+			sleep(10); // TODO for testing purposes
 			set_time_limit(self::TIME_LIMIT_PER_MAIL);
-			$email = $receiver->getAttribute('email');
-			$code = $receiver->getAttribute('unsubscribe-code');
+			$email = 'placeholder@test.test'; //$receiver->getAttribute('email'); TODO for testing purposes
+			$code = '123'; //$receiver->getAttribute('unsubscribe-code'); TODO for testing purposes
 			// include email for server record purposes
 			$linkToUnsubscribe = $this->path(self::PATH_UNSUBSCRIBE_CODE, ['address' => $email, 'code' => $code]);
 			$message = '<p><a href="' . $linkToMailing . '">' . __('ml-if-unreadable-use-link') . '</a></p>';
 			$message .= $mailing->getHtml();
 			$message .= '<p><a href="' . $linkToUnsubscribe . '">' . __('ml-unsubscribe') . '</a></p>';
-			$this->sendMail($mailing->getAttribute('subject'), $message, [$email], $senderData['email'], $senderData['name']);
+			//TODO disabled for testing purposes $this->sendMail($mailing->getAttribute('subject'), $message, [$email], $senderData['email'], $senderData['name']);
+
+			$this->xml->lockAndReload();
+			$mailing = $this->xml->getElementById($mailingId);
+			$mailing->setAttribute('progress', ++$count);
```

Comment:
Should the progress attribute be removed when the send is complete?

Reply:
Then the progress checker will find no final progress '10/10' and cannot determine that it has finished (using the way it has been build now).

Reply:
But the progress endpoint could just look at the status for that? Even more, if the endpoint is switched to JSON, it could maybe even return a "status" (or "completed") attributed as well?

Reply:
Yes, will do that :)

Reply:
And done